tests/nodescaling/scale_test.go

```diff
@@ -0,0 +1,165 @@
+		{"Scaling 2 nodes - etcd/cp roles per 1 node " + config.StandardClientName.String(), nodeRolesShared, scaleUpRolesShared, scaleDownRolesShared, r.standardUserClient},
+		{"Scaling 3 nodes - 1 role per node " + config.StandardClientName.String(), nodeRolesDedicated, scaleUpRolesDedicated, scaleDownRolesDedicated, r.standardUserClient},
+	}
+
+	for _, tt := range tests {
+		clusterConfig := *r.clusterConfig
+		clusterConfig.Nodepools = tt.nodeRoles
+		clusterConfig.ScaledUpNodepools = tt.scaleUpNodeRoles
+		clusterConfig.ScaledDownNodepools = tt.scaleDownNodeRoles
+
+		var scaledUpCount, scaledDownCount int64
+
+		for _, scaleUpNodepool := range tt.scaleUpNodeRoles {
+			scaledUpCount += scaleUpNodepool.Quantity
+		}
+
+		for _, scaleDownNodepool := range tt.scaleDownNodeRoles {
+			scaledDownCount += scaleDownNodepool.Quantity
+		}
+
+		clusterName := namegen.AppendRandomString(provisioning.TFP)
+
+		r.Run((tt.name), func() {
+			provisioning.Provision(r.T(), clusterName, r.terraformConfig, &clusterConfig, r.terraformOptions)
+			provisioning.VerifyCluster(r.T(), tt.client, clusterName, r.terraformConfig, r.terraformOptions, &clusterConfig)
+
+			provisioning.ScaleUp(r.T(), clusterName, r.terraformOptions, &clusterConfig)
+			provisioning.VerifyCluster(r.T(), tt.client, clusterName, r.terraformConfig, r.terraformOptions, &clusterConfig)
+			provisioning.VerifyNodeCount(r.T(), tt.client, clusterName, scaledUpCount)
+
+			provisioning.ScaleDown(r.T(), clusterName, r.terraformOptions, &clusterConfig)
+			provisioning.VerifyCluster(r.T(), tt.client, clusterName, r.terraformConfig, r.terraformOptions, &clusterConfig)
+			provisioning.VerifyNodeCount(r.T(), tt.client, clusterName, scaledDownCount)
+			cleanup.Cleanup(r.T(), r.terraformOptions)
+		})
+	}
+}
+
+func (r *ScaleTestSuite) TestScaleDynamicInput() {
+	tests := []struct {
+		name   string
+		client *rancher.Client
+	}{
+		{config.StandardClientName.String(), r.standardUserClient},
+	}
+
+	for _, tt := range tests {
+		clusterName := namegen.AppendRandomString(provisioning.TFP)
+
+		r.Run((tt.name), func() {
+			provisioning.Provision(r.T(), clusterName, r.terraformConfig, r.clusterConfig, r.terraformOptions)
+			provisioning.VerifyCluster(r.T(), r.client, clusterName, r.terraformConfig, r.terraformOptions, r.clusterConfig)
+
+			provisioning.ScaleUp(r.T(), clusterName, r.terraformOptions, r.clusterConfig)
+			provisioning.VerifyCluster(r.T(), r.client, clusterName, r.terraformConfig, r.terraformOptions, r.clusterConfig)
+			provisioning.VerifyNodeCount(r.T(), r.client, clusterName, r.clusterConfig.ScaledUpNodeCount)
+
+			provisioning.ScaleDown(r.T(), clusterName, r.terraformOptions, r.clusterConfig)
+			provisioning.VerifyCluster(r.T(), r.client, clusterName, r.terraformConfig, r.terraformOptions, r.clusterConfig)
+			provisioning.VerifyNodeCount(r.T(), r.client, clusterName, r.clusterConfig.ScaledDownNodeCount)
+			cleanup.Cleanup(r.T(), r.terraformOptions)
```

Comment:
I was curious about this but forgot to ask in meetings; why do we explicitly call the cleanup function? I thought testing.T is supposed to take care of that for us. If it needs to be added for each test, can we make sure this is in the parent README.md file for others to reference?

Reply:
This needs to be called each time because Terraform is not going to allow each test case to properly run in a set for loop like we have. Terraform by design is not going to fully cleanup resources unless you explicitly tell it to do so, so this has to be here.